### Credit Card Detection ###


In [ ]:
# importing the liberares
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px

import warnings
warnings.filterwarnings('ignore')


from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import GaussianNB , MultinomialNB , BernoulliNB
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import GradientBoostingClassifier
from lightgbm import LGBMClassifier


from sklearn.ensemble import AdaBoostClassifier, BaggingClassifier
from xgboost import XGBClassifier
from sklearn.pipeline import Pipeline
from sklearn.model_selection import cross_val_score
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report , precision_score

In [ ]:
# loading the dataset
df = pd.read_csv('/content/sample_data/creditcard.csv')
df.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [ ]:
# Data Information
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [ ]:
# checking for missing Values
df.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [ ]:
# removing the missing values
df = df.dropna()

In [ ]:
df.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [ ]:
df.describe()

Time            V1            V2            V3            V4  \
count  284807.000000  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean    94813.859575  1.168375e-15  3.416908e-16 -1.379537e-15  2.074095e-15   
std     47488.145955  1.958696e+00  1.651309e+00  1.516255e+00  1.415869e+00   
min         0.000000 -5.640751e+01 -7.271573e+01 -4.832559e+01 -5.683171e+00   
25%     54201.500000 -9.203734e-01 -5.985499e-01 -8.903648e-01 -8.486401e-01   
50%     84692.000000  1.810880e-02  6.548556e-02  1.798463e-01 -1.984653e-02   
75%    139320.500000  1.315642e+00  8.037239e-01  1.027196e+00  7.433413e-01   
max    172792.000000  2.454930e+00  2.205773e+01  9.382558e+00  1.687534e+01   

                 V5            V6            V7            V8            V9  \
count  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean   9.604066e-16  1.487313e-15 -5.556467e-16  1.213481e-16 -2.406331e-15   
std    1.380247e+00  1.332271e+00  1.237094e+00  1.194353e+00  1.098632e+00   
min   -1.137433e+02 -2.616051e+01 -4.355724e+01 -7.321672e+01 -1.343407e+01   
25%   -6.915971e-01 -7.682956e-01 -5.540759e-01 -2.086297e-01 -6.430976e-01   
50%   -5.433583e-02 -2.741871e-01  4.010308e-02  2.235804e-02 -5.142873e-02   
75%    6.119264e-01  3.985649e-01  5.704361e-01  3.273459e-01  5.971390e-01   
max    3.480167e+01  7.330163e+01  1.205895e+02  2.000721e+01  1.559499e+01   

       ...           V21           V22           V23           V24  \
count  ...  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean   ...  1.654067e-16 -3.568593e-16  2.578648e-16  4.473266e-15   
std    ...  7.345240e-01  7.257016e-01  6.244603e-01  6.056471e-01   
min    ... -3.483038e+01 -1.093314e+01 -4.480774e+01 -2.836627e+00   
25%    ... -2.283949e-01 -5.423504e-01 -1.618463e-01 -3.545861e-01   
50%    ... -2.945017e-02  6.781943e-03 -1.119293e-02  4.097606e-02   
75%    ...  1.863772e-01  5.285536e-01  1.476421e-01  4.395266e-01   
max    ...  2.720284e+01  1.050309e+01  2.252841e+01  4.584549e+00   

                V25           V26           V27           V28         Amount  \
count  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  284807.000000   
mean   5.340915e-16  1.683437e-15 -3.660091e-16 -1.227390e-16      88.349619   
std    5.212781e-01  4.822270e-01  4.036325e-01  3.300833e-01     250.120109   
min   -1.029540e+01 -2.604551e+00 -2.256568e+01 -1.543008e+01       0.000000   
25%   -3.171451e-01 -3.269839e-01 -7.083953e-02 -5.295979e-02       5.600000   
50%    1.659350e-02 -5.213911e-02  1.342146e-03  1.124383e-02      22.000000   
75%    3.507156e-01  2.409522e-01  9.104512e-02  7.827995e-02      77.165000   
max    7.519589e+00  3.517346e+00  3.161220e+01  3.384781e+01   25691.160000   

               Class  
count  284807.000000  
mean        0.001727  
std         0.041527  
min         0.000000  
25%         0.000000  
50%         0.000000  
75%         0.000000  
max         1.000000  

[8 rows x 31 columns]

In [ ]:
# distribution of the legit transcation and fraudulent transcation
df['Class'].value_counts()

Class
0    284315
1       492
Name: count, dtype: int64

### Data Is Highly Imbalance

In [ ]:
### sepearte the data into legit and fraud
legit = df[df.Class == 0]
fraud = df[df.Class == 1]

In [ ]:
print(legit.shape)
print(fraud.shape)

(284315, 31)
(492, 31)


In [ ]:
### measurement on the Amount Column in Legit
legit.Amount.describe()

count    284315.000000
mean         88.291022
std         250.105092
min           0.000000
25%           5.650000
50%          22.000000
75%          77.050000
max       25691.160000
Name: Amount, dtype: float64

In [ ]:
## measurement on the Amount Column in Fraud
fraud.Amount.describe()

count     492.000000
mean      122.211321
std       256.683288
min         0.000000
25%         1.000000
50%         9.250000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

In [ ]:
# compare the both the values
df.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94838.202258  0.008258 -0.006271  0.012171 -0.007860  0.005453   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.002419  0.009637 -0.000987  0.004467  ... -0.000644 -0.001235   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.000024  0.000070  0.000182 -0.000072 -0.000089 -0.000295 -0.000131   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       88.291022  
1      122.211321  

[2 rows x 30 columns]

- Number of Fraudulant Transcation : 284315
- Number of Legit Transcation : 492


In [ ]:
legit_sample = legit.sample(n=492)


In [ ]:
# concatenate the DataFrame
new_df = pd.concat([legit_sample, fraud], axis=0)
new_df.sample(5)

Time         V1        V2         V3        V4         V5  \
234632  148053.0   1.261324  2.726800  -5.435019  5.342759   1.447043   
156020  107041.0  -0.034575  0.196215   1.158813 -1.783633  -0.346253   
8866     12141.0  -2.520594  1.472805   0.616720 -0.927833  -0.704913   
34866    37860.0   1.137707 -0.717353  -0.063728 -0.552453  -0.903790   
42936    41305.0 -12.980943  6.720508 -13.455636  8.698610 -11.479552   

              V6         V7        V8        V9  ...       V21       V22  \
234632 -1.442584  -0.898702  0.123062 -2.748496  ...  0.209086 -0.425938   
156020 -0.286420  -0.074665  0.127710  2.543820  ... -0.210574 -0.273665   
8866   -0.902595   0.273930  0.275107  2.591406  ... -0.398975 -0.131186   
34866  -0.631256  -0.537038  0.093243 -0.910273  ...  0.081987 -0.145634   
42936  -2.681519 -14.019291  8.218191 -7.930900  ...  2.549628 -0.532228   

             V23       V24       V25       V26       V27       V28  Amount  \
234632 -0.154440 -0.018820  0.632234  0.192922  0.468181  0.280486    1.59   
156020  0.020214 -0.505546 -0.937119 -0.450820  0.195300  0.174922    6.95   
8866    0.094493  0.392750  0.146448  0.888378  0.573194  0.528543   45.50   
34866   0.003365  0.094359  0.190527 -0.343273 -0.005282  0.036879   99.00   
42936  -0.235096  0.673209  0.226598 -0.006168 -1.185696 -0.747361   59.68   

        Class  
234632      1  
156020      0  
8866        0  
34866       0  
42936       1  

[5 rows x 31 columns]

In [ ]:
new_df.shape

(984, 31)

In [ ]:
# Calculate Q1 (25th percentile) and Q3 (75th percentile)
Q1 = df.quantile(0.25)
Q3 = df.quantile(0.75)
IQR = Q3 - Q1

# Define outliers as values outside 1.5*IQR range
outliers = ((df < (Q1 - 1.5 * IQR)) | (df > (Q3 + 1.5 * IQR)))

# Melt DataFrame for Plotly visualization
df_melted = df.melt(var_name="Feature", value_name="Value")

# Plot using Plotly Boxplot
fig = px.box(df_melted, x="Feature", y="Value", title="Outlier Detection using Box Plot",
             points="outliers")  # Show outliers as individual points

fig.show()


In [ ]:
new_df['Class'].value_counts()

Class
0    492
1    492
Name: count, dtype: int64

### Splitting the DataSet in to X and y

In [ ]:
X = new_df.drop('Class', axis=1)
y = new_df['Class']

In [ ]:
X.shape

(984, 30)

In [ ]:
y.shape

(984,)

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42 , stratify=y)

### Model Training ###

In [ ]:
model = LogisticRegression()

model.fit(X_train, y_train)

LogisticRegression()

In [ ]:
y_pred = model.predict(X_test)

In [ ]:
print("accuracy_score", accuracy_score(y_test, y_pred))
# print("Precision_score", precision_score(y_test, y_pred))
print("confusion_matrix", confusion_matrix(y_test, y_pred))
print("classification_report", classification_report(y_test, y_pred))

accuracy_score 0.9593908629441624
confusion_matrix [[96  3]
 [ 5 93]]
classification_report               precision    recall  f1-score   support

           0       0.95      0.97      0.96        99
           1       0.97      0.95      0.96        98

    accuracy                           0.96       197
   macro avg       0.96      0.96      0.96       197
weighted avg       0.96      0.96      0.96       197



In [ ]:

models = {
    "Logistic Regression": LogisticRegression(max_iter=1000),
    "Random Forest": RandomForestClassifier(n_estimators=100),
    "SVM": SVC(),
    "Naive Bayes": GaussianNB(),
    "K-Nearest Neighbors": KNeighborsClassifier(n_neighbors=5),
    "Decision Tree": DecisionTreeClassifier(),
    "Gradient Boosting": GradientBoostingClassifier(),
    "XGBoost": XGBClassifier(),

    "AdaBoost": AdaBoostClassifier(),
    "Bagging": BaggingClassifier(),
    'logistic': LogisticRegression(),
    "Bernoulli Naive Bayes": BernoulliNB()
}

# Evaluate models
for name, model in models.items():
    if isinstance(model, (XGBClassifier, GradientBoostingClassifier)):
        # Train and evaluate without pipeline
        model.fit(X_train, y_train)
        y_pred = model.predict(X_test)
    else:
        steps = []
        if not isinstance(model, (GaussianNB, MultinomialNB, BernoulliNB)):
            steps.append(('scaler', StandardScaler()))
        steps.append(('classifier', model))
        pipeline = Pipeline(steps)
        pipeline.fit(X_train, y_train)
        y_pred = pipeline.predict(X_test)

    accuracy = accuracy_score(y_test, y_pred)



    print(f"{name} - Test Accuracy: {accuracy:.4f}")



Logistic Regression - Test Accuracy: 0.9442
Random Forest - Test Accuracy: 0.9340
SVM - Test Accuracy: 0.9340
Naive Bayes - Test Accuracy: 0.8579
K-Nearest Neighbors - Test Accuracy: 0.9137
Decision Tree - Test Accuracy: 0.9137
Gradient Boosting - Test Accuracy: 0.9340
XGBoost - Test Accuracy: 0.9492
AdaBoost - Test Accuracy: 0.9340
Bagging - Test Accuracy: 0.9289
logistic - Test Accuracy: 0.9442
Bernoulli Naive Bayes - Test Accuracy: 0.9086


In [ ]:
import numpy as np
import pandas as pd
import plotly.express as px
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split

# Generate synthetic data
np.random.seed(42)
X = np.random.rand(100, 1) * 10  # Feature (independent variable)
y = 2.5 * X + np.random.randn(100, 1) * 5  # Target with some noise

# Convert to DataFrame
df = pd.DataFrame({"X": X.flatten(), "y": y.flatten()})

# Train-test split
X_train, X_test, y_train, y_test = train_test_split(df["X"], df["y"], test_size=0.2, random_state=42)

# Train Linear Regression Model
model = LinearRegression()
model.fit(X_train.values.reshape(-1, 1), y_train)

# Generate predictions
df["Predicted"] = model.predict(df["X"].values.reshape(-1, 1))

# Plot using Plotly
fig = px.scatter(df, x="X", y="y", title="Linear Regression Fit with Plotly", labels={"X": "Feature", "y": "Target"})
fig.add_scatter(x=df["X"], y=df["Predicted"], mode="lines", name="Regression Line")

# Show interactive plot
fig.show()
